[FEATURE] Gérer un accès à durée spécifique, étendue et limitée, pour les comptes anonymes (en accès simplifié) (PIX-4633)

igorissen opened the pull request:
## :unicorn: Problème

Les utilisateurs qui jouent des parcours simplifiés voient apparaître des erreurs 403 au bout de 20 minutes, ce qui correspond à la durée de vie de l'access token. En effet, ils utilisent des comptes anonymes qui sont gérés côté Ember par l’authenticator `anonymous`, lequel ne gère pas les refresh tokens.

## :robot: Solution

- Ne pas faire de gestion via un refresh token.
- Créer un access token avec une durée de vie spécifique, limitée mais plus longue que la durée de vie par défaut des access tokens, via une nouvelle configuration dédiée uniquement à l'authentification d'un compte anonyme (valeur de 4h).

## :rainbow: Remarques

- En prod le problème a été contourné temporairement par l’allongement de la durée de vie des access tokens (passée à 4 heures au lieu de 20 minutes).
- Au moment de la mise en production de cette branche il faudra, pour la durée de vie par défaut des access tokens, remettre une valeur plus basse.

## :100: Pour tester

1. Allez sur [Parcours simplifié](https://app-pr4807.review.pix.fr/campagnes/SIMPLIFIE).
2. Ouvrir la console développeur et se mettre sur l'onglet `Network|Réseau`.
3. Cliquez sur le bouton `Je commence`.
4. Constatez dans la liste des appels réseaux qu'une ligne `anonymous` est présente (POST /api/token/anonymous) avec une réponse 200 et un access token dans le body de la requête.

en local pour tester l'invalidation de la session

1. Réduisez le temps d'expiration de `ANONYMOUS_ACCESS_TOKEN_LIFESPAN` a 10s
2. Redémarrer votre api
3. Allez sur [Parcours simplifié](https://app-pr4807.review.pix.fr/campagnes/SIMPLIFIE)
4. Cliquez sur le bouton `Je commence`.
5. Parcourez la campagne et attendez que le token expire
7. Remarquez que vous êtes redirigé vers la page de connexion une fois le token expiré

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4807.review.pix.fr
- App (.org): https://app-pr4807.review.pix.org
- Orga: https://orga-pr4807.review.pix.fr
- Certif: https://certif-pr4807.review.pix.fr
- Admin: https://admin-pr4807.review.pix.fr
- API: https://api-pr4807.review.pix.fr/api/

 Please check it out!

igorissen committed:
chore(api): update environment variables sample file

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>

igorissen committed:
feat(api): add new settings for anonymous authentication

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>

igorissen committed:
feat(api): add new method

create an access token for an anonymous user with his own expiration delay settings

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>
Co-authored-by: yschoueri <yasmina.schoueri@gmail.com>

igorissen committed:
refactor(api): use new method from token service

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>
Co-authored-by: yschoueri <yasmina.schoueri@gmail.com>

igorissen committed:
feat(api): add missing test

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>

igorissen committed:
feat(api): handle unauthorized request

with an error code SESSION_EXPIRED and redirect the user to the login page

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>
Co-authored-by: yschoueri <yasmina.schoueri@gmail.com>

igorissen committed:
feat(mon-pix): handle unauthorized request

with an error code SESSION_EXPIRED and redirect the user to the login page

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>
Co-authored-by: yschoueri <yasmina.schoueri@gmail.com>

yschoueri commented on api/lib/domain/services/token-service.js:
```diff
@@ -19,6 +19,12 @@ function createAccessTokenFromUser(userId, source) {
+function createAccessTokenFromAnonymousUser(userId) {
+  const expirationDelaySeconds = settings.anonymous.accessTokenLifespanMs / 1000;
+  const accessToken = _createAccessToken({ userId, source: 'pix', expirationDelaySeconds });
+  return { accessToken, expirationDelaySeconds };
```
`expirationDelaySeconds` n'est pas utilisé, on pourrait retourner l'accessToken directement

yschoueri commented on api/lib/domain/services/token-service.js:
```diff
@@ -19,6 +19,12 @@ function createAccessTokenFromUser(userId, source) {
 
+function createAccessTokenFromAnonymousUser(userId) {
+  const expirationDelaySeconds = settings.anonymous.accessTokenLifespanMs / 1000;
+  const accessToken = _createAccessToken({ userId, source: 'pix', expirationDelaySeconds });
+  return accessToken;
```
```suggestion
  return _createAccessToken({ userId, source: 'pix', expirationDelaySeconds });
```

lego-technix committed:
feat(api): add acceptance test

Co-authored-by: Marc-Aurèle Darche <marc-aurele.darche@pix.fr>
Co-authored-by: Ismaël Gorissen <ismael.gorissen@gmail.com>

Anne-Gaelle-S commented on high-level-tests/e2e/cypress/support/index.js:
```diff
@@ -26,4 +26,14 @@ beforeEach(() => {
     win.sessionStorage.clear();
   });
+
+  cy.on('uncaught:exception', (err) => {
```
**remarque**
J'imagine qu'il a du y avoir beaucoup de fun pour faire ces changements dans les tests e2e 😅

Anne-Gaelle-S commented on mon-pix/app/adapters/application.js:
```diff
@@ -35,4 +35,12 @@ export default class Application extends JSONAPIAdapter {
   }
+
+  handleResponse(status, headers, payload) {
+    if (status === 401 && payload.errors[0].code === 'SESSION_EXPIRED') {
+      return this.session.invalidate();
+    }
+
+    return super.handleResponse(...arguments);
+  }
```
🎉

Anne-Gaelle-S commented on mon-pix/tests/unit/adapters/application_test.js:
```diff
@@ -95,4 +96,42 @@ describe('Unit | Adapters | ApplicationAdapter', function () {
+
+        // then
+        sinon.assert.notCalled(session.invalidate);
+        sinon.assert.calledOnce(REST.prototype.handleResponse);
```
🥵

Anne-Gaelle-S reviewed (approved):
Pas encore testé fonctionnellement :)

Anne-Gaelle-S commented:
J'ai un petit soucis quand je met le `ANONYMOUS_ACCESS_TOKEN_LIFESPAN` à 10secondes . Je suis pas redirigé vers la page de login mais plutôt vers la page d'erreur : 
<img width="1671" alt="image" src="https://user-images.githubusercontent.com/38167520/187228868-427a5189-3bdb-49bf-970a-57574ece1b5d.png">
